### ISSUE TITLE
AttributeError: 'NoneType' object has no attribute 'get'

### ISSUE THREAD

**Issue by tttffff** [archlinux/archinstall#992]
When passing in `--disk_layout` with my admittedly non-standard layout, I am getting odd behavior. I can not replicate this when using the standard layout, but maybe it is a symptom of another bug. This is using the latest [commit on master](https://github.com/archlinux/archinstall/commit/62a6aec197b0e83ab91fdec1ab15427ecf749dac).

- If `"encryption": true` is set and nothing is entered for `Set encryption password` on the interactive screen, I am prompted to enter the encryption password after clicking `Install` and the installation is a success (this is expected behavior, and how I am installing Arch at the minute).
- If `"encryption": true` is set and something is entered into `Set encryption password` on the interactive screen, the below error message is given (not-expected behavior).
- If `"encrpytion": false` is set and nothing is entered for `Set encryption password` on the interactive screen, installation will complete without encryption of the drive (expected behavior).
- If `"encryption": false` is set and something is entered into `Set encryption password` on the interactive screen, the below error message is given (not-expected behavior).
- Not setting `"encryption"` in the disk_layout acts the same as setting it to false.

The error message:
```
 ! Formatting [BlockDevice(/dev/sda, size=32.0GB, free_space=3145kB+3146kB, bus_type=sata)] in 5....4....3....2....1....
Creating a new partition label on /dev/sda
Adding partition....
Adding partition to BlockDevice(/dev/sda, size=32.0GB, free_space=34.4GB, bus_type=sata), 3MiB->2048MiB
Formatting /dev/sda1 -> vfat
Marking partition Partition(path=/dev/sda1, size=2.0, PARTUUID=40e9df38-01, fs=vfat) as bootable.
Setting boot on on (parted) partition index 1
Adding partition....
Adding partition to BlockDevice(/dev/sda, size=32.0GB, free_space=3145kB+32.2GB, bus_type=sata), 2051MiB->100%
Encrypting Partition(path=/dev/sda2, size=30.0, PARTUUID=40e9df38-02, parent=/dev/sda, fs=crypto_LUKS) (This might take a while)
Formatting /dev/mapper/ainstsda2 -> btrfs
'NoneType' object has no attribute 'get'
Traceback (most recent call last):
  File "/root/kickassarchinstall/archinstall/archinstall/examples/guided.py", line 129, in perform_installation
    installation.mount_ordered_layout(archinstall.arguments['disk_layouts'])
  File "/root/kickassarchinstall/archinstall/archinstall/lib/installer.py", line 240, in mount_ordered_layout
    with (luks_handle := luks2(partition['device_instance'], loopdev, password, auto_unmount=False)) as unlocked_device:
  File "/root/kickassarchinstall/archinstall/archinstall/lib/luks.py", line 57, in __exit__
    raise args[1]
  File "/root/kickassarchinstall/archinstall/archinstall/lib/installer.py", line 241, in mount_ordered_layout
    if partition.get('generate-encryption-key-file',False) and not self._has_root(partition):
  File "/root/kickassarchinstall/archinstall/archinstall/lib/installer.py", line 211, in _has_root
    for mountpoint in [sub_list[subvolume] if isinstance(sub_list[subvolume],str) else sub_list[subvolume].get("mountpoint") for subvolume in sub_list]:
  File "/root/kickassarchinstall/archinstall/archinstall/lib/installer.py", line 211, in <listcomp>
    for mountpoint in [sub_list[subvolume] if isinstance(sub_list[subvolume],str) else sub_list[subvolume].get("mountpoint") for subvolume in sub_list]:
AttributeError: 'NoneType' object has no attribute 'get'

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/root/kickassarchinstall/archinstall/archinstall/__main__.py", line 15, in <module>
    archinstall.run_as_a_module()
  File "/root/kickassarchinstall/archinstall/./archinstall/__init__.py", line 272, in run_as_a_module
    script.execute()
  File "/root/kickassarchinstall/archinstall/archinstall/lib/profiles.py", line 194, in execute
    self.spec.loader.exec_module(sys.modules[self.namespace])
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/root/kickassarchinstall/archinstall/archinstall/examples/guided.py", line 275, in <module>
    perform_installation(archinstall.storage.get('MOUNT_POINT', '/mnt'))
  File "/root/kickassarchinstall/archinstall/archinstall/examples/guided.py", line 125, in perform_installation
    with archinstall.Installer(mountpoint, kernels=archinstall.arguments.get('kernels', ['linux'])) as installation:
  File "/root/kickassarchinstall/archinstall/archinstall/lib/installer.py", line 145, in __exit__
    print(_("[!] A log file has been created here: {} {}").format(os.path.join(storage['LOG_PATH'], storage['LOG_FILE'])))
  File "/root/kickassarchinstall/archinstall/archinstall/lib/translation.py", line 58, in format
    return self.message.format(*args)
IndexError: Replacement index 1 out of range for positional args tuple
```

My disk layout:
```
{
    "/dev/sda": {
        "partitions": [
            {
                "boot": true,
                "encrypted": false,
                "filesystem": {
                    "format": "fat32"
                },
                "mountpoint": "/boot",
                "size": "2048MiB",
                "start": "3MiB",
                "type": "primary",
                "wipe": true
            },
            {
                "btrfs": {
                    "subvolumes": {
                        "@": null,
                        "@/home": "/home",
                        "@/root": "/root",
                        "@/log": "/var/log",
                        "@/pkg": "/var/cache/pacman/pkg",
                        "@/vmimages": "/var/lib/libvrt/images",
                        "@/mariadb": "/var/lib/mariadb",
                        "@/mysql": "/var/lib/mysql",
                        "@/postgres": "/var/lib/pgsql",
                        "@/.snapshots": "/.snapshots",
                        "@/.snapshots/1/snapshot": "/"
                    }
                },
                "encrypted": true,
                "filesystem": {
                    "format": "btrfs"
                },
                "mountpoint": null,
                "size": "100%",
                "start": "2051MiB",
                "type": "primary",
                "wipe": true
            }
        ],
        "wipe": true
    }
}
```

**Comment by Torxed** [comment]
we'll have a look :) thanks for reporting with great detail!

**Comment by tttffff** [comment]
I should mention that the information above came from testing in VirtualBox. When I tried on bare metal, I got different errors about partprobe failing on /dev/sda (this is after it encrypted and formatted my drive, and mounted all the subvolumes). My drive is an NVME, and sda was actually the system of the USB that I booted from. For this reason, I got the idea to try loading into RAM (one of the options on the Arch installation ISO). When loaded into RAM, archinstall functions as my issue above outlines. I currently have a fully functional install on bare metal using my custom layout by: loading into RAM, setting —disk_layout “encryption”: true, and, pressing install before stating the passphrase for the drive. If you would like me to test further, please let me know, but it will be in a VM as I’ve not got a spare machine right now (plus I LOVE my current set up, think I’ve got my endgame Arch install here 😁)

**Comment by niluzz** [comment]
I'm having the same problem when installing on a partitioned Windows hard drive! I can't do it at all, even though I inform all the information necessary for installation! Erro “ nonetype object has no attribute “path”

**Comment by wllacer** [comment]
@tttffff Many Thanks for your feedback.  The cause of the bug is so obvious I don't know why it never happened here. And still doesn't happen to me
https://github.com/archlinux/archinstall/blob/86a21711b54ac69b11ce1bd5af130383c6a6a6bc/archinstall/lib/installer.py#L212-L216
You have a subvolume which will not  be mounted.  A patch will be forthcoming ASAPfindmnt
Your setup is really interesting. Could be enhanced with following code
```json
               "btrfs": {
                    "subvolumes": {
                        "@": null,
                        "@/home": "/home",
                        "@/root": "/root",
                        "@/log": "/var/log",
                        "@/pkg": "/var/cache/pacman/pkg",
                        "@/vmimages": {
                                "mountpoint":"/var/lib/libvrt/images",
                                "options":["nodatacow"]
                        },
                        "@/mariadb": {
                                "mountpoint":"/var/lib/mariadb",
                                "options":["nodatacow"]
                        },
                        "@/mysql": {
                                "mountpoint":"/var/lib/mysql",
                                "options":["nodatacow"]
                        },
                        "@/postgres": {
                                "mountpoint":"/var/lib/pgsql",
                                "options":["nodatacow"]
                        },
                        "@/.snapshots": "/.snapshots",
                        "@/.snapshots/1/snapshot": "/"
                    }
                },
 
```
if you want to use nodatacow on the data directories.

**Comment by wllacer** [comment]
@niluzz Your problem seems more related to issue #938. We would be happy if you could send there the files you'll find in the /var/archinstall/log of the machine were you tried the installation

**Comment by wllacer** [comment]
@tttffff Just a  question. Do you have a `generate-encryption-key-file` in the real disk_layout file? It happens in some old version files, and when it is present is the only way i've found / think is possible to replicate your error

**Comment by tttffff** [comment]
@wllacer thank you for looking into this and for you’re advise about my layout. I had no idea that I could set mount options in the disk_layout file, and it’s really helpful to know.

I don’t have`generate-encryption-key-file` in my disk-layout file, nor do I use a credential file, so it’s still a mystery why you can’t replicate the error without that.

My layout is a more OpenSUSE style, I picked it because I can use `snapper rollback $num` to rollback my system, it’s super convenient. After configuring snapper, I just get Grub to load from the default btrfs sub volume (by `sed -i 's/rootflags=subvol=${rootsubvol}//' /etc/grub.d/10_linux` and a quick `grub-mkconfig` then it’s good to go. This layout actually got discussed on issue [781](https://github.com/archlinux/archinstall/issues/781) and I think ShaunTheQuietGamer’s comment lays it out nicely (I did notice that you said that you don’t care for rollbacks though).

**Comment by Torxed** [comment]
I wanna clarify that we're not opposed to snapshotting and using tools to rollback.
We just did not want to add or fix that currently due to btrfs functionality didn't allow us - so we only cared about fixing that (first) :)

I personally would love snapshotting and rollback features, so I'm keeping a close eye on wllacer's progress as wllacer is the mastermind behind all the BTRFS fixes :D 

@wllacer correct me if I'm wrong, but this is solved entirely with #1004?
If not I can re-open!

**Comment by tttffff** [comment]
@Torxed, I should be able to test at the weekend and confirm (we just moved house and don't have the internet until Friday afternoon). I also want to try using the mount options in the disk_layout file like @wllacer explained. Thank you both for your everything, this is a great project.
